Add experiment_routes Blueprint on experiment class

alecpm opened the pull request:
## Description
This PR implements an `experiment_routes` Flask Blueprint on the `Experiment` class, similar to the optional `extra_routes` Flask Blueprint defined at the module level. It also updates the MCMCP and Sheep Market demos to use the new route registration mechanism and updates the docs to include the Blueprint as part of the `Experiment` class API. There is also a custom route registration decorator which makes it possible to define the route functions at the class level (though they must be either `classmethod` or plain functions and cannot access the experiment instance directly).

Additionally, to simplify Dashboard route registration this PR now includes a parallel `dashboard_tab` decorator which registers an Experiment class `classmethod` or function as a dashboard tab. This simplifies custom dashboard setup significantly, and provides similar benefits to the `experiment_route` decorator: allowing for inheritance/overrides, and preventing accidental double registration.

## Motivation and Context
See issue #2296 for motivation. The Blueprint has been renamed `experiment_routes` because I wanted to keep the option to use either or both module and class level Blueprints, and each registered Blueprint must have a unique name. Additionally, since a class level Blueprint seemed less that useful if the route functions could not be defined at the class level, I also added a decorator that allows functions and `classmethod` defined in the class to be used as routes.

The `dashboard_tab` decorator was added as a response to issues raised in [comments raised here](https://github.com/Dallinger/Dallinger/pull/2528#issuecomment-777056937).

## How Has This Been Tested?
I added an automated test that checks to see if a route registered on the experiment class is available to the experiment server. There are automated tests for the operation of the route decorators. I also updated the demos with custom routes to use the new method and tested them. Unfortunately, the custom route tests cannot be run in isolation because route registration happens during experiment_server import, and will generally only happen once during the test run, so the tests are marked as `xfail`. They should succeed if run in isolation using e.g. `pytest -k test_custom_route --runslow`.

alecpm commented:
Proposed CHANGELOG entry:

* Added `experiment_routes` Flask Blueprint to the `Experiment` class along with an experiment route registration decorator `dallinger.experiment.experiment_route`.

alecpm commented:
I'm attempting to do a little refactoring on this to make inheritance a little better and allow normal methods to be used, so hold off on reviews for now.

alecpm commented:
This is now ready for review with the updated implementation.

jessesnyder commented on dallinger/experiment_server/experiment_server.py:
```diff
@@ -86,6 +86,21 @@ def Experiment(args):
+except ImportError:
+    pass
+
 # Ideally, we'd only load recruiter routes if the recruiter is active, but
```
Might your decorator-based deferral solve the problem I encountered with recruiter-specific routes as well? I guess we'd need to switch to a model where we put recruiter classes in separate modules so we could import only the one we're going to use at run-time. Just want to keep this idea bubbling in the background while we're working on recruiter infrastructure.

alecpm replied:
It seems like you've got a workaround for these recruiter route issues, but if not a similar deferral like I did for experiment routes and dashboard routes could be useful there.

jessesnyder reviewed (commented):
This looks like a good solution to me. I want to spend a little more time with it tonight or tomorrow.

pmcharrison commented:
Unfortunately I won't be able to contribute much over the next few days because of an upcoming job interview, but I've asked Frank to fill in for me in terms of checking that everything works with our use cases.

alecpm commented:
@pmcharrison Related to your [dashboard related questions on PR 2528](https://github.com/Dallinger/Dallinger/pull/2528#issuecomment-777056937) I added a new decorator on this PR to make it easier and more reliable for experiments to add new dashboard tabs. It looks like this:

```
from dallinger.experiment_server.dashboard import dashboard_tab

class MyExperiment(Experiment):
    ...
    @dashboard_tab("My Tab Title", "my_tab_url")
    @classmethod
    def my_tab(cls):
        ... do some flask stuff ...
        return "Some flask page"
```

That decorator should take care of registering the tab with a title, decorating the function with `login_required`, and deferred registration of the route at experiment_server start. It works essentially the same as the `experiment_route` decorator from that PR, but with some extra bits for UI and authentication. There are tests for this functionality, but the integration tests are marked as `xfail` because they will generally fail when run as part of the full test suite because route registration only happens when the flask app is first loaded which currently occurs at import time, and can't be reset during test runs without restructuring the way Dallinger creates the Flask app.

alecpm commented on demos/dlgr/demos/sheep_market/experiment.py:
```diff
@@ -27,24 +27,17 @@ def create_network(self):
-    except TemplateNotFound:
-        abort(404)
+    @experiment_route("/drawings")
+    def getdrawings():
```
@jessesnyder If you'd prefer I can change the demos to use `@classmethod`, since I think that is best practice.

alecpm replied:
These use clasmethod now.

jessesnyder reviewed (approved):
There's a lot of complex stuff here, but it's carefully implemented and generally well documented. I left a few notes on things that I still found confusing.

alecpm commented on dallinger/experiment.py:
```diff
@@ -161,6 +182,13 @@ def __init__(self, session=None):
         else:
             self.widget = module.ExperimentWidget(self)
 
+    def extra_parameters(self):
```
I think this is an artifact of the fact that this was based on an earlier version of #2528 that had been merged to master. I'll see what I can do to rebase that away, as it shouldn't be here at all.

alecpm replied:
Rebase worked.

alecpm commented on dallinger/experiment.py:
```diff
@@ -1263,14 +1286,18 @@ def is_experiment_class(cls):
 
 
-def load():
+def load(initialize=True):
```
I think I can get rid of this actually. It was a workaround for issues that I believe I've resolved another way.

alecpm replied:
It was unused and is now removed.

alecpm commented on dallinger/experiment_server/experiment_server.py:
```diff
@@ -86,6 +86,43 @@ def Experiment(args):
     app.register_blueprint(extra_routes)
 
+
+try:
```
I think this is better now. The only ImportError I'm worried about is in `experiment.load()`.

alecpm commented on dallinger/experiment.py:
```diff
@@ -1307,3 +1334,34 @@ def load():
+        "kwargs": tuple(kwargs.items()),
+    }
+
+    def new_func(func):
```
I've refactored out the shared function, it's a little weird because it's a sort of meta-meta-decorator, but it avoids all the duplicated code I think.

alecpm commented on dallinger/experiment_server/experiment_server.py:
```diff
@@ -86,6 +86,45 @@ def Experiment(args):
+
+# Skipping coverage testing on this for now because it only runs at import time
+# and cannot be exercised within tests
+try:  # pragma: no cover
```
I've isolated the try/except to just the `experiment.load` call which is the only thing I worry might throw an `ImportError` now. I think that should suffice.

pmcharrison reviewed (changes requested):
Thanks a lot, @alecpm! I tested it and it is working well for me. I made a suggestion about the `dashboard_tab` functionality that might be worth checking out, let me know what you think. 

During the process of testing I discovered a bug in `extra_files` and pushed a one-line fix to it to the current merge request. Previously it generated nonsense when used to import directories. I hope this is OK, wasn't sure it was worth the overhead of a whole merge request.

alecpm commented:
@pmcharrison I've rebased this branch onto master and fixed the `method` argument. Additionally, I've removed the `rule` argument from the `dashboard_tab` decorator since it served little purpose. I changed the registration mechanism to be a little more foolproof/generic, and eliminated the `*args` which didn't seem to serve any useful purpose. Finally, I unified the implementation of this and the scheduled task decorator so they share a bit more code.

The `rule` argument is still in place for the more generice `experiment_route` decorator, since it allows for routes that use the same `rule` but e.g. different `methods` using different underlying functions.

It should be ready for re-test/review.

pmcharrison commented on dallinger/experiment_server/experiment_server.py:
```diff
@@ -86,6 +86,62 @@ def Experiment(args):
+            route_name = route["func_name"]
+            dashboard.dashboard.add_url_rule(
+                "/" + route_name,
+                end_point=route_name,
```
This should be `endpoint` - the code doesn't run if you use `end_point`.

pmcharrison reviewed (changes requested):
Hi @alecpm, worked well, except for two things:

- The `end_point` vs `endpoint` typo mentioned in the comments
- The rebase seems to have nuked my fix to the `extra_files` functionality. I have to wrap up for the day, but basically I just replaced this line in `deployment.py`:

``` py
dst_fileparts = [dst, filename] + dirnames + [fn]
```

with this line:

```py 
dst_fileparts = [dst, filename] + [os.path.relpath(dirpath, src)] + [fn]
```

The previous version didn't make much sense, I think maybe there was a mixup about what values `os.path.walk` returns?

Do you want to simply include this fix in the present pull request, and move forward? Or will you create a separate pull request?

alecpm commented:
The `end_point` thing is fixed now (oops). Not sure what happened with `dst_fileparts` but it looks like this branch is the same as `master` as far as that line goes, which means that some other merge to master undid that change. I'll make that change once I determine where/when it got changed.

alecpm commented:
The `dst_fileparts` thing is confusing me because the version on this branch has been on master since June 2020:

https://github.com/Dallinger/Dallinger/commit/666b059e36a3364a301cc4bbd74df6ce9e43b480

@pmcharrison Do you know where this fix was made? Was it only local to your setup before?

pmcharrison commented:
I think this is consistent with my understanding: so far we only used the
file version of this feature, not the directory version, so we never
encountered the bug.

I pushed the fix to this same branch, and you can see a reference to this
where it says an automated message like “Peter had trouble with continuous
integration”, but when I looked later it seemed that the commit
disappeared. Strange!

On Thu, 6 May 2021 at 11:02 pm, Alec Mitchell ***@***.***>
wrote:

> The dst_fileparts thing is confusing me because the version on this
> branch has been on master since June 2020:
>
> 666b059
> <https://github.com/Dallinger/Dallinger/commit/666b059e36a3364a301cc4bbd74df6ce9e43b480>
>
> @pmcharrison <https://github.com/pmcharrison> Do you know where this fix
> was made? Was it only local to your setup before?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Dallinger/Dallinger/pull/2525#issuecomment-833863589>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFNCKRT4NUUL3WXVPK2FVGDTML7VFANCNFSM4WE3LVWA>
> .
>

alecpm commented:
OK, I just added back your change. Hopefully codecov doesn't freak out again, otherwise I'll probably need to disable it temporarily.

alecpm commented:
I think this just needs approval and merge now.

pmcharrison reviewed (approved):
Thanks very much @alecpm !
